platform/include/transport_interface.h

```diff
@@ -19,21 +19,144 @@
+ * external transport layer protocol. The transport interface is implemented in
+ * @ref transport_interface.h. This interface allows the application layer
+ * protocol library to send and receive data over the transport layer. This
+ * interface does not handle connection and disconnection to the server of
+ * interest. The connection to the server of interest must be made before the
+ * library is used with @ref TransportInterface_t. The application is also
+ * responsible for disconnection from the server. This interface does not handle
+ * socket timeouts or any setup of TLS; these things must be done before using
+ * the library's functions that need to send and receive data.
+ * <br>
+ *
+ * The functions that must be implemented are:<br>
+ * - [Transport Receive](@ref TransportRecv_t)
+ * - [Transport Send](@ref TransportSend_t)
+ *
+ * Each of the functions above take in an opaque context @ref NetworkContext_t.
+ * The functions above and the context are also grouped together in the
+ * @ref TransportInterface_t structure:<br><br>
+ * @snippet this define_transportinterface
+ * <br>
+ *
+ * The following steps give guidance on implementing the transport interface:
+ *
+ * -# Implementing the @ref NetworkContext_t<br><br>
+ * @snippet this define_networkcontext
+ * <br>
+ * @ref NetworkContext_t is the incomplete type <b>struct NetworkContext</b>.
+ * The implemented struct NetworkContext should contain all of the information
+ * that is needed to receive and send data with the @ref TransportRecv_t
+ * and the @ref TransportSend_t implementations.<br>
+ * struct NetworkContext is typically implemented with the TCP socket context
+ * and a TLS context.<br><br>
+ * <b>Example code:</b>
+ * @code{c}
+ * struct NetworkContext
+ * {
+ *     struct MyTCPSocketContext tcpSocketContext;
+ *     struct MyTLSContext tlsContext;
+ * };
+ * @endcode
+ * <br>
+ * -# Implementing @ref TransportRecv_t<br><br>
+ * @snippet this define_transportrecv
+ * <br>
+ * The library expects this callback to populate a buffer with bytes from the
+ * network. Connection with the server of interest should have been made before
+ * invoking any library routines that receive data from the network.
+ * The @ref TransportRecv_t is typically implemented by calling directly the
+ * TLS layer function to receive data. If one is transferring over plaintext TCP,
+ * without TLS, then a call is made directly to the TCP layer function to receive
+ * data. @ref TransportRecv_t may be invoked multiple times by the library if
+ * fewer bytes than were requested to receive are returned.
+ * <br><br>
+ * <b>Example code:</b>
+ * @code{c}
+ * int32_t myNetworkRecvImplementation( const NetworkContext_t * pNetworkContext,
+ *                                      void * pBuffer,
+ *                                      size_t bytesToRecv )
+ * {
+ *     int32_t bytesReceived = 0;
+ *     bytesReceived = mySocketRecv( pNetworkContext->tcpSocketContext,
```

Comment:
The sample struct is provided for TLS over TCP, should this be TLSRecv?

Reply:
yes that makes more sense